main/distance_sensor.cpp

```diff
@@ -0,0 +1,160 @@
+#include "distance_sensor.h"
+
+#include <misc_utils/angle.h>
+
+#include <Eigen/Dense>
+
+#include <limits>
+#include <numbers>
+#include <ranges>
+
+#include <Arduino.h>
+#include <SparkFun_I2C_Mux_Arduino_Library.h>
+#include <SparkFun_VL53L1X.h>
+#include <esp_log.h>
+
+using namespace micromouse;
+
+static constexpr auto max_sensor_range = 0.15_m /* 1.3_m / 4 */;
+static constexpr auto max_predict_range = max_sensor_range /* 3.0_m */;
+
+// Sensors disposition relative to the center of rotation in meters.
+static constexpr std::array sensor_disposition{
+    std::pair(XCoord{62.0e-3f}, YCoord{54.0e-3f}),
+    std::pair(XCoord{90.0e-3f}, YCoord{40.0e-3f}),
+    std::pair(XCoord{94.0e-3f}, YCoord{0.0f}),
+    std::pair(XCoord{92.0e-3f}, YCoord{-39.0e-3f}),
+    std::pair(XCoord{67.0e-3f}, YCoord{-54.0e-3f}),
+};
+
+// Sensor angles relative to the robot in radians.
+static constexpr std::array sensor_angles{
+    Angle{std::numbers::pi_v<float> / 2},
+    Angle{0.703765f /* std::numbers::pi_v<float> / 4 */},
+    Angle{0.0f},
+    Angle{-0.649649f /* std::numbers::pi_v<float> / 4 */},
+    Angle{-std::numbers::pi_v<float> / 2},
+};
+static_assert(sensor_angles.size() == DistanceSensors::sensor_count);
+
+DistanceSensors &DistanceSensors::get_instance() noexcept
+{
+    static QWIICMUX mux;
+    static SFEVL53L1X current_distance_sensor;
+    static constinit DistanceSensors instance(mux, current_distance_sensor);
+    return instance;
+}
+
+void DistanceSensors::init(TwoWire &i2c, Vl53l1cdTimingBudget timing_budget) noexcept
+{
+    if (!m_mux.begin(QWIIC_MUX_DEFAULT_ADDRESS, i2c))
+    {
+        printf("Mux failed to begin. Please check wiring. Freezing...\n");
+        while (true)
+        {
+            delay(1);
+        }
```

Comment:
Do you know what would be cool here?

```suggestion
        led::infinite_loop([] { ESP_LOGE("sensor", "Mux failed to begin. Please check wiring. Freezing..."); });
```